Firmware/tmc2130.cpp

```diff
@@ -379,56 +393,84 @@ void tmc2130_check_overtemp()
+		} s;
+		uint32_t dw;
+	} chopconf;
+	static_assert(sizeof(ChopConfU::S) == 4);
+	static_assert(sizeof(ChopConfU) == 4);
+
+	chopconf.dw = 0; // Zero initialise
+	
+	chopconf.s.intpol = (mres != 0); // intpol to 256 only if microsteps aren't 256
 #ifdef TMC2130_DEDGE_STEPPING
-	uint8_t dedge = 1;
+	chopconf.s.dedge = 1;
 #else
-	uint8_t dedge = 0;
+	chopconf.s.dedge = 0;
 #endif
-	uint8_t toff = tmc2130_chopper_config[axis].toff; // toff = 3 (fchop = 27.778kHz)
-	uint8_t hstrt = tmc2130_chopper_config[axis].hstr; //initial 4, modified to 5
-	uint8_t hend = tmc2130_chopper_config[axis].hend; //original value = 1
-	uint8_t fd3 = 0;
-	uint8_t rndtf = 0; //random off time
-	uint8_t chm = 0; //spreadCycle
-	uint8_t tbl = tmc2130_chopper_config[axis].tbl; //blanking time, original value = 2
+	chopconf.s.toff = tmc2130_chopper_config[axis].toff; // toff = 3 (fchop = 27.778kHz)
+	chopconf.s.hstrt = tmc2130_chopper_config[axis].hstr; // initial 4, modified to 5
+	chopconf.s.hend = tmc2130_chopper_config[axis].hend; // original value = 1
+	chopconf.s.fd = 0;
+	chopconf.s.disfdcc = 0;
+	chopconf.s.rndtf = 0; //random off time
+	chopconf.s.chm = 0; //spreadCycle
+	chopconf.s.tbl = tmc2130_chopper_config[axis].tbl; //blanking time, original value = 2
+	chopconf.s.vsense = 0;
+	chopconf.s.vhighfs = 0;
+	chopconf.s.vhighchm = 0;
+	chopconf.s.sync = 0;
+	chopconf.s.mres = mres;
 	if (axis == E_AXIS)
 	{
 #if defined(TMC2130_INTPOL_E) && (TMC2130_INTPOL_E == 0)
-        intpol = 0;
+		chopconf.s.intpol = 0;
 #endif
 #ifdef TMC2130_CNSTOFF_E
-		// fd = 0 (slow decay only)
-		hstrt = 0; //fd0..2
-		fd3 = 0; //fd3
-		hend = 0; //sine wave offset
-		chm = 1; // constant off time mod
+		chopconf.s.hstrt = 0; // fd0..2
+		chopconf.s.fd = 0; // fd3
+		chopconf.s.hend = 0; // sine wave offset
+		chopconf.s.chm = 0; // constant off time mod
 #endif //TMC2130_CNSTOFF_E
-//		toff = TMC2130_TOFF_E; // toff = 3-5
-//		rndtf = 1;
 	}
 #if defined(TMC2130_INTPOL_XY) && (TMC2130_INTPOL_XY == 0)
     else if (axis == X_AXIS || axis == Y_AXIS) {
-        intpol = 0;
+        chopconf.s.intpol = 0;
     }
 #endif
 #if defined(TMC2130_INTPOL_Z) && (TMC2130_INTPOL_Z == 0)
     else if (axis == Z_AXIS) {
-        intpol = 0;
+        chopconf.s.intpol = 0;
     }
 #endif
-//	DBG(_n("tmc2130_setup_chopper(axis=%d, mres=%d, curh=%d, curr=%d\n"), axis, mres, current_h, current_r);
-//	DBG(_n(" toff=%d, hstr=%d, hend=%d, tbl=%d\n"), toff, hstrt, hend, tbl);
 	if (current_r <= 31)
 	{
-		tmc2130_wr_CHOPCONF(axis, toff, hstrt, hend, fd3, 0, rndtf, chm, tbl, 1, 0, 0, 0, mres, intpol, dedge, 0);
-		tmc2130_wr(axis, TMC2130_REG_IHOLD_IRUN, 0x000f0000 | ((current_r & 0x1f) << 8) | (current_h & 0x1f));
-	}
-	else
-	{
-		tmc2130_wr_CHOPCONF(axis, toff, hstrt, hend, fd3, 0, rndtf, chm, tbl, 0, 0, 0, 0, mres, intpol, dedge, 0);
-		tmc2130_wr(axis, TMC2130_REG_IHOLD_IRUN, 0x000f0000 | (((current_r >> 1) & 0x1f) << 8) | ((current_h >> 1) & 0x1f));
+		chopconf.s.vsense = 1;
```

Comment:
this logic can be also hidden in the constructor of `chopconf` ... I think we have it like that in the MMU code

Reply:
I'm thinking about adding this `struct` to the MK3 firmware:

```cpp
struct MotorCurrents {
    bool vSense; ///< VSense current scaling
    uint8_t iRun; ///< Running current
    uint8_t iHold; ///< Holding current

    constexpr MotorCurrents(uint8_t ir, uint8_t ih)
        : vSense((ir < 32) ? 1 : 0)
        , iRun((ir < 32) ? ir : (ir >> 1))
        , iHold((ir < 32) ? ih : (ih >> 1)) {}
};
```
It will calculate the vsense flag for us. But since both Run and Hold currents can be updated separately at run time we need to make sure the Vsense bit is always re-calculated.

At the cost of 4 bytes of SRAM, we can save ~50 bytes of Flash by create a global array of these structs for each axis. It also seems to make the code cleaner in general. I'm not ready to push it just yet, but it's maybe one way to resolve your comment :)

Reply:
@leptun quick question about `iHold((ir < 32) ? ih : (ih >> 1)) {}`

I wonder about the scenario where both `iHold, iRun = 31`, but what if the user suddenly changed `iRun = 33`, does `iHold` need to be updated with the >> bit shift?

Reply:
Yes. vsense gets applied to both IHOLD and IRUN, so IHOLD also needs to be shifted if IRUN got shifted. It is enough to check ir<32 for both fields since always IRUN >= IHOLD.

Reply:
`IRUN >= IHOLD` this was also something I remember you mentioned about the MMU FW. I don't think we are currently enforcing this.

Reply:
Yeah, it might be a good idea to constrain that value, but I would skip that if it doesn't get optimized well.

Reply:
@DRracer I've refactored this code now, should be much better now 👍